### PR TITLE
Exercises bundle no. 7, 8, 9, 10, 11, 12

### PR DESCRIPTION
**Pull request by purplehepatica**

### CODE REVIEW

File: Podstawy/7/7.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            return num
+
+
+def task4():
+    def nums_multiply(*args):
+        nums_multiplication = 1
+
+        for num in args:
+            if not isinstance(num, int):
+                continue
```
Comment (oskarissimus) [inline Podstawy/7/7.py:51]: Good job in making additional validation :)

File: Podstawy/7/7.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+        shortest_list = even_nums if len(even_nums) < len(odd_nums) else odd_nums
+
+        longest_list = even_nums if len(even_nums) > len(odd_nums) else odd_nums
+
+        for i in range(len(shortest_list)):
+            even_odd_nums_chain.append(shortest_list[i])
+            even_odd_nums_chain.append(longest_list[i])
+
+        even_odd_nums_chain.extend(longest_list[len(shortest_list):])
```
Comment (oskarissimus) [inline Podstawy/7/7.py:71]: technically correct, but too much parentheses nesting for my taste
I would recommend using temp var like `leftover` or smth. Code with less nested parentheses tend to be more readable.
however it might use more memory (not sure) so It's good to be midful about that
```py
leftover = longest_list[len(shortest_list):]
even_odd_nums_chain.extend(leftover)
```

File: Podstawy/7/7.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+def task6():
+    nums = [1, 5, 23, 45, 67, 101, 126, 179, 1001, 1999]
+
+    def get_two_digit_nums(nums_list):
+
+        two_digit_nums = []
+
+        for num in nums_list:
+            if len(str(num)) == 2:
```
Comment (oskarissimus) [inline Podstawy/7/7.py:86]: technically correct, and readable, but more it seems that it would be more performant if You would use smth like `10 <= num <= 99`

File: Podstawy/7/7.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+def task7():
+    def insert_start_fuel_level():
+        return int(input("Wprowadź wartość paliwa między 5 000, a 30 000 litrów: "))
+
+    def is_start_fuel_level_valid(level):
+        return 5000 < level < 30000
+
+    def insert_astronauts_num():
+        return int(
+            input("Wprowadź liczbę astronautów na pokładzie, która jest liczbą dodatnią i nie większą od 7: "))
```
Comment (oskarissimus) [inline Podstawy/7/7.py:103]: it looks like formatting isn't quite right. might be useful to setup "format on save" in Your IDE

File: Podstawy/7/7.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    def is_astronauts_num_valid(astronauts_num):
+        return 0 < astronauts_num <= 7
+
+    def measure_distance(fuel_level, astronauts, distance):
+        while fuel_level >= 0:
+            print(f'Przebyto: {distance} kilometrów.')
+            print(f'Posiadamy {fuel_level} litrów paliwa.')
+
+            distance += 100
+            fuel_level -= 300 + 100 * astronauts
```
Comment (oskarissimus) [inline Podstawy/7/7.py:114]: it's a matter of taste, but I use `+=`, `-=`, etc only with single number. Whenever more terms are used I find such code confusing (I have to think how order of operation works) and it is easier to make a bug

File: Podstawy/10/1-9.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
+import json
+
+
+def task1():
+    # Określ, czy poniższe ścieżki do plików.txt są względne / bezwzględne:
+    # - C:\przyklad.txt -> ścieżka bezwzględna
+    # - \katalog\przyklad.txt -> ścieżka względna
+    # - C:\outer_dir\inner_dir\przyklad.txt -> ścieżka bezwzględna
+    # - \outer_dir\innedr_dir\przyklad.txt -> ścieżka względna
```
Comment (oskarissimus) [inline Podstawy/10/1-9.py:9]: wszystkie te ścieżki są bezwzględne
na windowsie nie jestem ekspertem w ścieżkach, ale aby się o tym przekonać można wejść w CMD lub powershell i użyć `cd` lub `ls`
na linuksie sprawa jest prosta 
-> ścieżka zaczyna się od `/` -> bezwzględna
-> ścieżka zaczyna się od czegokolwiek innego -> względna

File: Podstawy/10/1-9.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+def task3():
+    with open("./przyklad.txt", "r") as file:
+
+        lines = file.readlines()
+
+        # Parzyste linijki będą nieparzystymi indeksami na liście
+        for i in range(len(lines)):
+            if i % 2 != 0:
+                print(lines[i])
```
Comment (oskarissimus) [inline Podstawy/10/1-9.py:31]: I would recommend to discover other solutions - using slicing

File: Podstawy/10/1-9.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        words = line.strip().split(" ")
+        cleared_words = []
+
+        for i in range(len(words)):
+            if i == 0 or words[i] != words[i - 1]:
+                cleared_words.append(words[i])
+
+        cleared_text_lines.append(" ".join(cleared_words))
+
+    cleared_text = " \n".join(cleared_text_lines)
```
Comment (oskarissimus) [inline Podstawy/10/1-9.py:75]: typo?
```suggestion
    cleared_text = "\n".join(cleared_text_lines)
```

File: Podstawy/10/1-9.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+            reversed_dict[value] = key
+
+        return reversed_dict
+
+    with open("./output.json", "w", encoding="utf-8") as file:
+        new_dict = reverse_key_value(klucz1="wartość1", klucz2="wartość2", klucz3="wartość3")
+
+        dict_json = json.dumps(new_dict, indent=4, ensure_ascii=False)
+
+        file.write(dict_json)
```
Comment (oskarissimus) [inline Podstawy/10/1-9.py:96]: there is `json.dump()` that writes directly to a file

File: Podstawy/10/1-9.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        file.write(cleared_text)
+
+
+def task8():
+    def reverse_key_value(**kwargs):
+        reversed_dict = {}
+
+        for item in kwargs.items():
+            key, value = item
+            reversed_dict[value] = key
```
Comment (oskarissimus) [inline Podstawy/10/1-9.py:87]: you can also destructure directly in for statement

File: Podstawy/10/1-9.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+def task9():
+    packets = {}
+
+    print("Interface Status")
+    print("================================================================================")
+    print("DN                                                 Description           Speed    MTU  ")
+    print("-------------------------------------------------- --------------------  ------  ------")
+
+    with open("./data.json", "r") as file:
+        packets = json.loads(file.read())["imdata"]
```
Comment (oskarissimus) [inline Podstawy/10/1-9.py:108]: could as well use json.load

File: Podstawy/10/1-9.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+    for packet in packets:
+        attributes = packet["l1PhysIf"]["attributes"]
+
+        dn = attributes["dn"]
+
+        if len(attributes["dn"]) == 41:
+            dn += "  "
+        elif len(attributes["dn"]) == 42:
+            dn += " "
```
Comment (oskarissimus) [inline Podstawy/10/1-9.py:118]: there is also `str.ljust()` to achieve similar things
https://docs.python.org/3/library/stdtypes.html#str.ljust
or string formatting https://docs.python.org/3/library/string.html#format-examples
```py
print(f'{dn:<43}')
# or
dn.ljust(43)
```

File: Podstawy/11/1-8.py [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+            commission = self.count_commission(amount)
+
+            self.balance += amount
+            self.charge_commission(commission)
+
+        # 3.
+        def withdraw(self, amount):
+            if amount > self.balance:
+                print("Brak wystarczających środków!")
+                return
```
Comment (oskarissimus) [inline Podstawy/11/1-8.py:77]: good job with early return in validation

File: Podstawy/11/1-8.py [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+
+    class Deck:
+        def __init__(self, cards):
+            self.cards = cards
+
+        def shuffle(self):
+            random.shuffle(self.cards)
+
+        def deal(self):
+            card = cards.pop(-1)
```
Comment (oskarissimus) [inline Podstawy/11/1-8.py:117]: pop() by default returns last element from the list, so this param could be omitted, I guess You added this for sake of expliciteness?

File: Podstawy/11/1-8.py [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+
+    class Card:
+        def __init__(self, value, figure):
+            self.value = value
+            self.figure = figure
+
+    for figure in figures:
+        for value in values:
+            card = Card(value, figure)
+            cards.append(card)
```
Comment (oskarissimus) [inline Podstawy/11/1-8.py:128]: this is all good, but as interesting note - itertools product might here be used as alternative
https://docs.python.org/3/library/itertools.html#itertools.product

File: Podstawy/11/1-8.py [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+    class Order:
+        def __init__(self, id, name, price):
+            self.id = id
+            self.name = name
+            self.price = price
+
+    pluszak = Order(121, "Miś Amisz", 100)
+    owad = Order(3301, "Cykada", 7)
+
+    zamowienia = Manager()
```
Comment (oskarissimus) [inline Podstawy/11/1-8.py:166]: why polish var names?

File: Podstawy/11/1-8.py [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+    class Manager:
+        def __init__(self):
+            self.orders = {}
+
+        def add(self, order, amount=1):
+            if self.orders.get(order):
+                self.orders[order] += amount
+                return
+
+            self.orders[order] = amount
```
Comment (oskarissimus) [inline Podstawy/11/1-8.py:144]: You might wanna take a look at defaultdict, to solve these kinds of problems in future :)

File: Podstawy/11/1-8.py [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+
+    class Case:
+        def __init__(self, case):
+
+            self.name = case["name"]
+            self.created_task = case["created_task"]
+            self.end_task = case["end_task"]
+
+        def retrieve_seconds(self):
+            date_pattern = "%Y-%m-%dT%H:%M:%S+00:00"
```
Comment (oskarissimus) [inline Podstawy/11/1-8.py:250]: its all good, but this pattern looks like an iso datetime format? not sure, but it might be simpler (last part, after `+` sign is timezone offset)

File: Podstawy/11/1-8.py [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+
+            created_date = datetime.strptime(self.created_task, date_pattern)
+
+            end_date = 0
+            if self.end_task is None:
+                end_date = datetime.today()
+            else:
+                end_date = datetime.strptime(self.end_task, date_pattern)
+
+            total_seconds = (end_date - created_date).total_seconds()
```
Comment (oskarissimus) [inline Podstawy/11/1-8.py:260]: nice

File: Podstawy/12/2.py [hunk 20]
```diff
@@ -0,0 +1,105 @@
+class Depot:
+    def __init__(self, name):
+        self.name = name
+        self.type = "Zwykła zajezdnia"
```
Comment (oskarissimus) [inline Podstawy/12/2.py:4]: I guess we could remove this property, since it is redundant information

File: Podstawy/12/2.py [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+        self.vehicles = []
+
+    def __str__(self):
+        info = [
+            f"Nazwa zajezdni: {self.name}",
+            f"Typ: {self.type}",
```
Comment (oskarissimus) [inline Podstawy/12/2.py:10]: here You would somehow extract type of depot from class

File: Podstawy/12/2.py [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+        return super().__str__() + f"\n-\nŁączna liczba wagonów tramwajowych: {total_wagons_num}"
+
+
+class Vehicle:
+    def __init__(self, max_speed, number, depot):
+        self.max_speed = max_speed
+        self.number = number
+        self.depot = depot.name
+
+        depot.add(self)
```
Comment (oskarissimus) [inline Podstawy/12/2.py:57]: this line feels dangerous. I understand that it is kinda mentioned in the task (every vehicle should know its depot and every depot should know its vehicles)
**HOWEVER**
caling `.add()` on depot has 1 one major problem - it mutates param passed to constructor. generally speaking we should avoid mutuating parameters of functions if possible. If it is necessary to mutate parameter it should be obvious to the caller that it happens (eg indicated by name). others say that you should return some value or mutate parameter, not both.
To recap:
1. I would suggest to keep constructors to minimum (just initialize attrs), it keeps the code nice and predictable.
2. dont mutate parameters of any function, if it is not super-clear for caller (by appropriate func name, or param name)
it will help You to avoid hard-to-debug problems in future

File: Podstawy/12/2.py [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+        return super().__str__() + f"\nLiczba wagonów: {self.wagons}"
+
+
+def main():
+    zajezdnia_autobusowa = BusDepot("Grodzka")
+    zajezdnia_tramwajowa = TramDepot("Mickiewicza")
+
+    autobus = Bus(200, 111, zajezdnia_autobusowa, 1000)
+    tramwaj1 = Tram(240, 23, zajezdnia_tramwajowa, 3)
+    tramwaj2 = Tram(220, 24, zajezdnia_tramwajowa, 1)
```
Comment (oskarissimus) [inline Podstawy/12/2.py:98]: why polish var names?

File: Podstawy/12/4.py [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+    def __init__(self):
+        self.items = []
+
+    def add(self, item):
+        self.items.append(item)
+
+    def show(self):
+        print("---\nZapisane:")
+        for i in range(len(self.items)):
+            print(f"{i + 1}. {self.items[i]}")
```
Comment (oskarissimus) [inline Podstawy/12/4.py:78]: here, enumerate would be more pythonic

File: Podstawy/12/4.py [hunk 25]
```diff
@@ -0,0 +1,99 @@
...
+
+
+class NotesSubManager(SubManager):
+    def __init__(self):
+        super().__init__()
+
+
+class CardsSubManager(SubManager):
+    def __init__(self):
+        super().__init__()
```
Comment (oskarissimus) [inline Podstawy/12/4.py:89]: It seems like it is not actually needed?

Comment (oskarissimus) [reply Podstawy/12/4.py:89]: You might use `pass` if You want to subclass without changing

File: Podstawy/8/1-7.py [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+    for i in range(1, 5):
+        print(i * "*")
+
+    for i in range(5, 0, -1):
+        print(i * "*")
+
+
+def task2():
+    def check_if_palindrome(word):
+        is_palindrome = 'jest' if word.lower() == word.lower()[::-1] else 'nie jest'
```
Comment (oskarissimus) [inline Podstawy/8/1-7.py:11]: it is correct, however I think there is too  much happening in this single line. I think that code seems to be more readable if we have just one thing per line
so anyway I would break this line into two

File: Podstawy/8/1-7.py [hunk 27]
```diff
@@ -0,0 +1,94 @@
...
+        print("---\n")
+
+
+def task6():
+    def num_frequency(num=None):
+        if num is None or isinstance(num, str):
+            print("Nie wybrano liczby!")
+            quit()
+
+        unique_digits = list(set(map(int, str(num))))
```
Comment (oskarissimus) [inline Podstawy/8/1-7.py:63]: This line is making me feel ambivalent. on the one hand there is waaay to much nesting, but on the other it just looks so neat. you got me here. I like it but i cant recomend doing this

File: Podstawy/8/1-7.py [hunk 28]
```diff
@@ -0,0 +1,94 @@
...
+    fibonacci_nums = [0, 1]
+    penultimate_num = fibonacci_nums[0]
+    last_num = fibonacci_nums[1]
+
+    for i in range(1, num + 1):
+
+        current_num = penultimate_num + last_num
+
+        if num == 1 or current_num > num:
+            break
```
Comment (oskarissimus) [inline Podstawy/8/1-7.py:88]: do we actually need this condition?

File: Podstawy/9/1-6.py [hunk 29]
```diff
@@ -0,0 +1,135 @@
...
+
+
+def task2():
+    import random
+
+    nums = set()
+
+    def rm_even_nums(num):
+        if num % 2 != 0:
+            return num
```
Comment (oskarissimus) [inline Podstawy/9/1-6.py:18]: filter function should return boolean not value or None (aka `Optional[int]`)
in this case it would work the same way (because of every number from this set would be coerced to True)
but, if test set would have `0` in it, it would not be correct

File: Podstawy/9/1-6.py [hunk 30]
```diff
@@ -0,0 +1,135 @@
...
+def task4():
+    cities_rainfalls = {}
+
+    while True:
+        choice = input("Wprowadź miasto i opady oddzielone spacją: ")
+
+        if choice == "":
+            break
+
+        city, rainfall = choice.split(" ")
```
Comment (oskarissimus) [inline Podstawy/9/1-6.py:52]: nice

File: Podstawy/9/1-6.py [hunk 31]
```diff
@@ -0,0 +1,135 @@
...
+
+        if cities_rainfalls.get(city):
+            cities_rainfalls[city] += int(rainfall)
+            continue
```
Comment (oskarissimus) [inline Podstawy/9/1-6.py:56]: i feel like else would be better choice here? unsure, but continue is "breaking" flow of a program. I would find this more readable with else statement instead

File: Podstawy/9/1-6.py [hunk 32]
```diff
@@ -0,0 +1,135 @@
...
+
+        cities_rainfalls[city] = int(rainfall)
+
+    for city in cities_rainfalls:
+        print(f"{city} : {cities_rainfalls[city]}")
```
Comment (oskarissimus) [inline Podstawy/9/1-6.py:61]: good usage of iteration over a dict.
in this case it might make sense to iterate over `cities_rainfalls.values()`
but either way is fine

File: Podstawy/9/1-6.py [hunk 33]
```diff
@@ -0,0 +1,135 @@
...
+        ['Rosie', 'Garlic Bread', 4.35],
+        ['Rosie', 'Veneziana', 9.40],
+        ['Rosie', 'Tiramisu', 4.90],
+    ]
+
+    bill = {}
+
+    for bill_item in bill_items:
+
+        person, dish, price = bill_item
```
Comment (oskarissimus) [inline Podstawy/9/1-6.py:84]: this is all fine, but You might also destructure right in for statement

Comment (oskarissimus) [reply Podstawy/9/1-6.py:84]: like so:
```py
for person, dish, price in bill_items:
```

File: Podstawy/9/1-6.py [hunk 34]
```diff
@@ -0,0 +1,135 @@
...
+
+        if not bill.get(person):
+            bill[person] = {}
+            bill[person]["potrawy"] = []
+            bill[person]["cena"] = 0
```
Comment (oskarissimus) [inline Podstawy/9/1-6.py:89]: defaultdict come in hady in such cases

File: Podstawy/9/1-6.py [hunk 35]
```diff
@@ -0,0 +1,135 @@
...
+                if isinstance(value, list):
+                    for subvalue in value:
+                        values.append(subvalue)
+                else:
+                    values.append(value)
+
+        if isinstance(data[key], dict):
+            for subkey in data[key]:
+                for value in data[key][subkey]:
+                    values.append(value)
```
Comment (oskarissimus) [inline Podstawy/9/1-6.py:127]: quite nice impl. for traversing leaves of this object tree, one might also use some recursive approach or for example some queue approach (either lifo or fifo)

**Review by oskarissimus — APPROVED** [review]
very nice and readable code overall. I nitpicked some things, but the one important is **mutation of a param in class constructor** other than that, great work and great pace :)
